File: seminar/src/main/java/com/server/dosopt/seminar/service/PostService.java
package com.server.dosopt.seminar.service;

import com.server.dosopt.seminar.domain.Member;
import com.server.dosopt.seminar.domain.Post;
import com.server.dosopt.seminar.dto.request.post.PostCreateRequest;
import com.server.dosopt.seminar.dto.request.post.PostUpdateRequest;
import com.server.dosopt.seminar.dto.response.post.PostGetResponse;
import com.server.dosopt.seminar.repository.MemberRepository;
import com.server.dosopt.seminar.repository.PostJpaRepository;
import jakarta.persistence.EntityNotFoundException;
import java.util.List;
import java.util.NoSuchElementException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class PostService {

    private final PostJpaRepository postRepository;
    private final MemberRepository memberRepository;  // 연관관계 매핑이 되어 있는 Member의 레포지토리도 주입받아서 사용

    @Transactional
    public String create(PostCreateRequest reqeust, Long memberId) {
        Member member = memberRepository.findByIdOrThrow(memberId);
        Post post = postRepository.save(
                Post.builder()
                        .member(member)
                        .title(reqeust.title())
                        .content(reqeust.content()).build());
        return post.getPostId().toString();

    }

    public PostGetResponse getById(Long postId) {
        Post post = findPostById(postId);
        return PostGetResponse.of(post);
    }

    public List<PostGetResponse> getPosts(Long memberId) {
        return postRepository.findAllByMemberId(memberId).stream()
                .map(post -> PostGetResponse.of(post))
                .toList();
    }

    @Transactional
    public void update(PostUpdateRequest request, Long postId) {
        Post post = findPostById(postId);
        post.updateContent(request.content());   // update 이후 save 하지 않아도 반영된다!!
    }

    @Transactional
    public void delete(Long postId) {
        postRepository.deleteById(postId);
    }

    private Post findPostById(Long postId) {
        return postRepository.findById(postId)
                .orElseThrow(() -> new EntityNotFoundException("존재하지 않는 게시글입니다."));
    }


}

File: kotlin-seminar/src/main/kotlin/sopt/org/kotlinseminar/service/PostService.kt
package sopt.org.kotlinseminar.service

import jakarta.persistence.EntityNotFoundException
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import sopt.org.kotlinseminar.domain.Post
import sopt.org.kotlinseminar.dto.request.post.PostCreateRequest
import sopt.org.kotlinseminar.dto.request.post.PostUpdateRequest
import sopt.org.kotlinseminar.dto.response.post.PostGetResponse
import sopt.org.kotlinseminar.repository.MemberRepository
import sopt.org.kotlinseminar.repository.PostJpaRepository

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
class PostService(
        private val postRepository: PostJpaRepository,
        private val memberRepository: MemberRepository
) {

    @Transactional
    fun create(request: PostCreateRequest, memberId: Long): String {
        val member = memberRepository.findByIdOrThrow(memberId)
        val post = postRepository.save(
                Post(member = member,
                    title = request.title,
                    content = request.content,
                    imageUrl = null)
        )
        return post.id.toString()
    }

    fun getById(postId: Long): PostGetResponse {
        val post = findPostById(postId)
        return PostGetResponse.of(post);
    }

    fun getPosts(memberId: Long): List<PostGetResponse> {
        return postRepository.findAllByMemberId(memberId).stream()
                .map { PostGetResponse.of(it) }
                .toList()
    }

    @Transactional
    fun update(request: PostUpdateRequest, postId: Long) {
        val post = findPostById(postId)
        post.updateContent(request.content)
    }

    @Transactional
    fun delete(postId: Long) {
        postRepository.deleteById(postId)
    }

    private fun findPostById(postId: Long): Post = postRepository.findById(postId)
            .orElseThrow { EntityNotFoundException("존재하지 않는 게시물입니다.") }


}